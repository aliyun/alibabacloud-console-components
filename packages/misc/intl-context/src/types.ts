/**
 * @public
 */
export interface IMessages {
  [key: string]: string
}

/**
 * @public
 */
export interface IIntlCtxValue {
  locale?: string
  messages?: IMessages
  // Provider allow users to extend IIntlCtxValue with anything
  [key: string]: any
}

/**
 * @public
 */
export interface IProviderProps {
  locale?: string
  messages?: IMessages
  /**
   * User can add new properties to provide.
   */
  extend?:
    | object
    | ((providerProps: IProviderProps, contextValue: IIntlCtxValue) => object)
  // Provider allow users to pass any props
  // and use them in 'extend(this.props, contextValue)'
  [key: string]: any
}
