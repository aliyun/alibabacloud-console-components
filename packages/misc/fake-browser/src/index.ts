export { default } from './FakeBrowser'
export { default as FakeBrowserWithWrapper } from './WithWrapper'
