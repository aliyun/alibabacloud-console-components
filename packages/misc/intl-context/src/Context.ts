import { createContext } from 'react'
import { IIntlCtxValue } from './types'

export default createContext<IIntlCtxValue>({})
