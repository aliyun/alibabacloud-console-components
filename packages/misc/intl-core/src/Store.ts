const glob: any = (() => {
  if (typeof window !== 'undefined') return window
  if (typeof global !== 'undefined') return global
  return {}
})()

const WIND_INTL_STORE = '__WIND_INTL_STORE__'

if (!glob[WIND_INTL_STORE]) {
  glob[WIND_INTL_STORE] = {}
}

class Store {
  private _store: { [key: string]: any }

  public static create(isCrossModules = true) {
    return new Store(isCrossModules)
  }

  public constructor(isCrossModules = true) {
    this._store = isCrossModules ? glob[WIND_INTL_STORE] : {}
  }

  public set(key: string, value: any) {
    this._store[key] = value
  }

  public get(key: string) {
    return this._store[key]
  }
}

export default Store
