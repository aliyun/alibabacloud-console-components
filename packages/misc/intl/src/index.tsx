import type {
  IWindIntlPublic,
  IWindIntlExtended,
} from '@alicloud/console-components-intl-core'
import { createReactIntlFromCfg } from './factory'

export type {
  ExtendIntl,
  IWindIntlPublic,
} from '@alicloud/console-components-intl-core'

const __intl = createReactIntlFromCfg()
/** @public */
const { IntlProvider } = __intl
/** @public */
const { withProvider } = __intl
/** @public */
const { Consumer } = __intl

/**
 * @public
 * This instance is created in top level code. And shared by
 * `import intl from '@alicloud/console-components-intl'`.
 * You can create your own instance with `reactIntlFactory`.
 */
const intlPublicTyped: IWindIntlPublic = __intl

export default intlPublicTyped
export { intlPublicTyped as intl, IntlProvider, withProvider, Consumer }
export {
  createReactIntlFromCfg,
  createReactIntlFromCfg as reactIntlFactory,
  createReactIntlFromInstance,
} from './factory'

export {
  ReactIntl,
  IntlBase,
  VanillaIntl,
  presets,
  withRcIntl,
} from '@alicloud/console-components-intl-core'
