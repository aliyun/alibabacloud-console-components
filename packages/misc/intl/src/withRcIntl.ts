/**
 * 兼容旧版rc-intl的目录结构
 */
export { withRcIntl as default } from './index'
