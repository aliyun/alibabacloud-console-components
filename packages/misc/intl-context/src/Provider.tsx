import React, { Component } from 'react'
import PropTypes from 'prop-types'
import isPlainObject from 'is-plain-object'
import Context from './Context'
import { IIntlCtxValue, IProviderProps } from './types'

/* eslint-disable react/static-property-placement */

/**
 * @public
 * Map the contextValue get from intl-context
 * into a new contextValue, and re-provide it to it's subtree.
 */
class Provider extends Component<IProviderProps> {
  public static propTypes = {
    locale: PropTypes.string,
    messages: PropTypes.objectOf(PropTypes.any),
    extend: PropTypes.oneOfType([
      PropTypes.objectOf(PropTypes.any),
      PropTypes.func,
    ]),
    children: PropTypes.node,
  }

  /**
   * User can add new properties to provide.
   */
  public computeExtendValue(contextValue: IIntlCtxValue): IIntlCtxValue {
    const { extend } = this.props

    if (!extend) {
      return {}
    }

    if (typeof extend === 'function') {
      return extend(this.props, contextValue)
    }

    if (isPlainObject(extend)) {
      return extend
    }

    return {}
  }

  public computeExactValueFrom(contextValue: IIntlCtxValue): IIntlCtxValue {
    const {
      locale: contextLocale = undefined,
      messages: contextMessages = undefined,
      ...restContextValue
    } = contextValue || {}

    const { locale, messages } = this.props

    return {
      ...restContextValue,
      locale: locale || contextLocale,
      messages: messages || contextMessages,
      ...this.computeExtendValue(contextValue),
    }
  }

  public render() {
    const { children } = this.props

    return (
      <Context.Consumer>
        {value => (
          <Context.Provider value={this.computeExactValueFrom(value)}>
            {children}
          </Context.Provider>
        )}
      </Context.Consumer>
    )
  }
}

export default Provider
