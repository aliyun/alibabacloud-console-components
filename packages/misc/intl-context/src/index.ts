import Provider from './Provider'
import Consumer from './Consumer'
import Context from './Context'

export { Provider, Consumer, Context }
export * from './types'
