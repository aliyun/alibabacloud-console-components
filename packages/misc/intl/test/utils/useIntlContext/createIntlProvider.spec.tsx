import React from 'react'
import { mount } from 'enzyme'

import { Consumer } from '@alicloud/console-components-intl-context'
import createIntlProvider from '../../../src/utils/useIntlContext/createIntlProvider'
import { ReactIntl } from '@alicloud/console-components-intl-core'
import { create } from '../../../src/factory'

describe('createIntlProvider', () => {
  let intlInstance = create(new ReactIntl())
  let { IntlProvider } = createIntlProvider(intlInstance)
  const TestConsumer: React.FC<any> = () => null
  beforeEach(() => {
    intlInstance = create(
      new ReactIntl({
        locale: 'locale-from-intlInstance',
        messages: { fromIntlInstance: 'message from intlInstance' },
      })
    )
    IntlProvider = createIntlProvider(intlInstance).IntlProvider
  })
  it("should get intl config from IWindIntlPublic if it's not in props", () => {
    const wrapper = mount(
      <IntlProvider>
        <Consumer>{value => <TestConsumer {...value} />}</Consumer>
      </IntlProvider>
    )
    expect(wrapper).toMatchSnapshot()
  })
  it('intl config from props should have higher priority than intl config from IWindIntlPublic', () => {
    const wrapper = mount(
      <IntlProvider
        locale="locale-from-props"
        messages={{ fromProps: 'message from props' }}
      >
        <Consumer>{value => <TestConsumer {...value} />}</Consumer>
      </IntlProvider>
    )
    expect(wrapper).toMatchSnapshot()
  })
  it('should provide flat rawMessages and nested messages', () => {
    const wrapper = mount(
      <IntlProvider
        locale="locale-from-props"
        messages={{
          'a.b.c.d': 'nested message from props 1',
          'a.b.e': 'nested message from props 2',
          'b.c.e': 'nested message from props 3',
          flat: 'flat message from props',
        }}
      >
        <Consumer>{value => <TestConsumer {...value} />}</Consumer>
      </IntlProvider>
    )
    expect(wrapper).toMatchSnapshot()
  })
})
