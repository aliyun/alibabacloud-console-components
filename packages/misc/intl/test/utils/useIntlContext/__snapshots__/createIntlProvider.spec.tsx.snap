// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createIntlProvider intl config from props should have higher priority than intl config from IWindIntlPublic 1`] = `
<IntlProvider
  baseComponentKeyPrefix="@wind_v2.base"
  locale="locale-from-props"
  messages={
    Object {
      "fromProps": "message from props",
    }
  }
>
  <Provider
    extend={[Function]}
    locale="locale-from-props"
    messages={
      Object {
        "fromProps": "message from props",
      }
    }
  >
    <ConfigProvider
      errorBoundary={false}
      locale={
        Object {
          "momentLocale": "locale-from-props",
        }
      }
      warning={true}
    >
      <TestConsumer
        intl={[Function]}
        locale="locale-from-props"
        messages={
          Object {
            "fromProps": "message from props",
          }
        }
        rawMessages={
          Object {
            "fromProps": "message from props",
          }
        }
      />
    </ConfigProvider>
  </Provider>
</IntlProvider>
`;

exports[`createIntlProvider should get intl config from IWindIntlPublic if it's not in props 1`] = `
<IntlProvider
  baseComponentKeyPrefix="@wind_v2.base"
>
  <Provider
    extend={[Function]}
    locale="locale-from-intlInstance"
    messages={
      Object {
        "fromIntlInstance": "message from intlInstance",
      }
    }
  >
    <ConfigProvider
      errorBoundary={false}
      locale={
        Object {
          "momentLocale": "locale-from-intlInstance",
        }
      }
      warning={true}
    >
      <TestConsumer
        intl={[Function]}
        locale="locale-from-intlInstance"
        messages={
          Object {
            "fromIntlInstance": "message from intlInstance",
          }
        }
        rawMessages={
          Object {
            "fromIntlInstance": "message from intlInstance",
          }
        }
      />
    </ConfigProvider>
  </Provider>
</IntlProvider>
`;

exports[`createIntlProvider should provide flat rawMessages and nested messages 1`] = `
<IntlProvider
  baseComponentKeyPrefix="@wind_v2.base"
  locale="locale-from-props"
  messages={
    Object {
      "a.b.c.d": "nested message from props 1",
      "a.b.e": "nested message from props 2",
      "b.c.e": "nested message from props 3",
      "flat": "flat message from props",
    }
  }
>
  <Provider
    extend={[Function]}
    locale="locale-from-props"
    messages={
      Object {
        "a": Object {
          "b": Object {
            "c": Object {
              "d": "nested message from props 1",
            },
            "e": "nested message from props 2",
          },
        },
        "b": Object {
          "c": Object {
            "e": "nested message from props 3",
          },
        },
        "flat": "flat message from props",
      }
    }
  >
    <ConfigProvider
      errorBoundary={false}
      locale={
        Object {
          "momentLocale": "locale-from-props",
        }
      }
      warning={true}
    >
      <TestConsumer
        intl={[Function]}
        locale="locale-from-props"
        messages={
          Object {
            "a": Object {
              "b": Object {
                "c": Object {
                  "d": "nested message from props 1",
                },
                "e": "nested message from props 2",
              },
            },
            "b": Object {
              "c": Object {
                "e": "nested message from props 3",
              },
            },
            "flat": "flat message from props",
          }
        }
        rawMessages={
          Object {
            "a.b.c.d": "nested message from props 1",
            "a.b.e": "nested message from props 2",
            "b.c.e": "nested message from props 3",
            "flat": "flat message from props",
          }
        }
      />
    </ConfigProvider>
  </Provider>
</IntlProvider>
`;
